File: API-Hook/api_hook.py
"""
API Hook Implementation - Python Example
用於監控與記錄 API 呼叫的 Python 實作範例
"""

import json
import time
import hashlib
import logging
from datetime import datetime
from typing import Dict, Any, Optional, Callable
from functools import wraps
import uuid

# 配置日誌
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger('api_hook')


class APIHook:
    """
    API Hook 核心類別
    提供 API 呼叫攔截、記錄與監控功能
    """
    
    def __init__(self, config_path: str = "config.yaml"):
        """
        初始化 API Hook
        
        Args:
            config_path: 配置檔案路徑
        """
        self.config = self._load_config(config_path)
        self.sensitive_fields = self.config.get('security', {}).get('sensitive_fields', [])
        logger.info("API Hook initialized")
    
    def _load_config(self, config_path: str) -> Dict[str, Any]:
        """載入配置檔案"""
        # 簡化實作：實際應從 YAML 檔案載入
        return {
            'security': {
                'sensitive_fields': ['password', 'api_key', 'token', 'secret'],
                'encrypt_sensitive_data': True,
                'hash_algorithm': 'SHA256'
            },
            'logging': {
                'level': 'INFO',
                'storage': 'database'
            },
            'performance': {
                'async_processing': True
            }
        }
    
    def _sanitize_data(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """
        清理敏感資料
        將敏感欄位替換為雜湊值或遮罩
        
        Args:
            data: 原始資料
            
        Returns:
            清理後的資料
        """
        sanitized = data.copy()
        
        for field in self.sensitive_fields:
            if field in sanitized:
                # 使用雜湊值替代敏感資料
                value = str(sanitized[field])
                hash_value = hashlib.sha256(value.encode()).hexdigest()[:16]
                sanitized[field] = f"[REDACTED:{hash_value}]"
        
        return sanitized
    
    def _create_log_entry(
        self,
        request_id: str,
        user_id: str,
        source_ip: str,
        method: str,
        endpoint: str,
        parameters: Dict[str, Any],
        response_code: int,
        response_time_ms: float,
        result: str,
        error_message: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        創建日誌條目
        
        Returns:
            格式化的日誌條目
        """
        return {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'request_id': request_id,
            'user_id': user_id,
            'source_ip': source_ip,
            'method': method,
            'endpoint': endpoint,
            'parameters': self._sanitize_data(parameters),
            'response_code': response_code,
            'response_time_ms': round(response_time_ms, 2),
            'security_context': {
                'authentication_method': 'OAuth2',
                'authorization_level': 'User',
                'session_id': request_id
            },
            'result': result,
            'error_message': error_message
        }
    
    def _save_log(self, log_entry: Dict[str, Any]):
        """
        儲存日誌
        實際實作應根據配置寫入資料庫、檔案或 Elasticsearch
        
        Args:
            log_entry: 日誌條目
        """
        # 簡化實作：僅輸出到標準輸出
        logger.info(f"API Log: {json.dumps(log_entry, ensure_ascii=False)}")
        
        # 實際實作範例：
        # if self.config['logging']['storage'] == 'database':
        #     self.db.insert('api_logs', log_entry)
        # elif self.config['logging']['storage'] == 'elasticsearch':
        #     self.es.index('api-logs', log_entry)
    
    def monitor(
        self,
        endpoint: str,
        security_level: str = "medium",
        log_params: bool = True,
        log_response: bool = False
    ) -> Callable:
        """
        裝飾器：監控 API 端點
        
        Args:
            endpoint: API 端點路徑
            security_level: 安全等級 (low, medium, high, critical)
            log_params: 是否記錄參數
            log_response: 是否記錄回應內容
            
        Returns:
            裝飾器函數
            
        Example:
            @hook.monitor(endpoint="/api/v1/users", security_level="high")
            def create_user(request):
                return {"status": "success"}
        """
        def decorator(func: Callable) -> Callable:
            @wraps(func)
            def wrapper(*args, **kwargs):
                request_id = str(uuid.uuid4())
                start_time = time.time()
                
                # 提取請求資訊（簡化，實際應從框架中提取）
                user_id = kwargs.get('user_id', 'anonymous')
                source_ip = kwargs.get('source_ip', '0.0.0.0')
                method = kwargs.get('method', 'GET')
                parameters = kwargs if log_params else {}
                
                try:
                    # 執行原始函數
                    result = func(*args, **kwargs)
                    response_code = 200
                    status = 'success'
                    error_message = None
                    
                except Exception as e:
                    result = {'error': str(e)}
                    response_code = 500
                    status = 'error'
                    error_message = str(e)
                    logger.error(f"API Error: {endpoint} - {str(e)}")
                    raise
                
                finally:
                    # 計算執行時間
                    response_time_ms = (time.time() - start_time) * 1000
                    
                    # 創建並儲存日誌
                    log_entry = self._create_log_entry(
                        request_id=request_id,
                        user_id=user_id,
                        source_ip=source_ip,
                        method=method,
                        endpoint=endpoint,
                        parameters=parameters,
                        response_code=response_code,
                        response_time_ms=response_time_ms,
                        result=status,
                        error_message=error_message
                    )
                    
                    self._save_log(log_entry)
                    
                    # 檢查異常行為
                    self._check_anomalies(log_entry)
                
                return result
            
            return wrapper
        return decorator
    
    def _check_anomalies(self, log_entry: Dict[str, Any]):
        """
        檢查異常行為
        
        Args:
            log_entry: 日誌條目
        """
        # 檢查回應時間
        if log_entry['response_time_ms'] > 5000:
            self._trigger_alert(
                'Slow Response Time',
                f"API {log_entry['endpoint']} took {log_entry['response_time_ms']}ms",
                'WARNING'
            )
        
        # 檢查錯誤
        if log_entry['result'] == 'error':
            self._trigger_alert(
                'API Error',
                f"API {log_entry['endpoint']} failed: {log_entry['error_message']}",
                'ERROR'
            )
        
        # 檢查未授權存取
        if log_entry['response_code'] == 401:
            self._trigger_alert(
                'Unauthorized Access',
                f"Unauthorized access attempt from {log_entry['source_ip']}",
                'CRITICAL'
            )
    
    def _trigger_alert(self, alert_name: str, message: str, severity: str):
        """
        觸發警報
        
        Args:
            alert_name: 警報名稱
            message: 警報訊息
            severity: 嚴重程度
        """
        alert = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'alert_name': alert_name,
            'message': message,
            'severity': severity
        }
        
        logger.warning(f"ALERT: {json.dumps(alert, ensure_ascii=False)}")
        
        # 實際實作：發送到警報系統、Email、Webhook 等
        # self._send_webhook(alert)
        # self._send_email(alert)
    
    def log_event(
        self,
        event_type: str,
        severity: str,
        message: str,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """
        手動記錄事件
        
        Args:
            event_type: 事件類型 (security, performance, audit, etc.)
            severity: 嚴重程度 (info, warning, error, critical)
            message: 事件訊息
            metadata: 額外的元資料
            
        Example:
            hook.log_event(
                event_type="security",
                severity="critical",
                message="Brute force attack detected",
                metadata={"ip": "192.168.1.100", "attempts": 10}
            )
        """
        event = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'event_type': event_type,
            'severity': severity,
            'message': message,
            'metadata': metadata or {}
        }
        
        logger.info(f"Event: {json.dumps(event, ensure_ascii=False)}")
        
        # 儲存事件
        self._save_log(event)
        
        # 如果是嚴重事件，觸發警報
        if severity in ['error', 'critical']:
            self._trigger_alert(event_type.title(), message, severity.upper())


# 使用範例
if __name__ == '__main__':
    # 初始化 API Hook
    hook = APIHook(config_path="config.yaml")
    
    # 範例 1：使用裝飾器監控 API 函數
    @hook.monitor(
        endpoint="/api/v1/users",
        security_level="high",
        log_params=True
    )
    def create_user(user_id: str, username: str, password: str, **kwargs):
        """創建使用者"""
        # 模擬 API 邏輯
        print(f"Creating user: {username}")
        return {
            'status': 'success',
            'user_id': '12345',
            'username': username
        }
    
    # 範例 2：使用裝飾器監控認證 API
    @hook.monitor(
        endpoint="/api/v1/auth/login",
        security_level="critical",
        log_params=True
    )
    def login(user_id: str, username: str, password: str, **kwargs):
        """使用者登入"""
        print(f"User login: {username}")
        if password == "wrong":
            raise Exception("Invalid credentials")
        return {
            'status': 'success',
            'token': 'jwt-token-here'
        }
    
    # 測試 API 呼叫
    print("\n=== 測試 1: 成功的 API 呼叫 ===")
    result = create_user(
        user_id='user123',
        username='john_doe',
        password='secure_password',
        source_ip='192.168.1.100',
        method='POST'
    )
    print(f"Result: {result}\n")
    
    # 測試失敗的 API 呼叫
    print("=== 測試 2: 失敗的 API 呼叫 ===")
    try:
        result = login(
            user_id='user456',
            username='jane_doe',
            password='wrong',
            source_ip='192.168.1.101',
            method='POST'
        )
    except Exception as e:
        print(f"Error caught: {e}\n")
    
    # 測試手動事件記錄
    print("=== 測試 3: 手動記錄安全事件 ===")
    hook.log_event(
        event_type="security",
        severity="critical",
        message="Suspicious login pattern detected",
        metadata={
            "ip": "192.168.1.100",
            "attempts": 10,
            "user": "admin"
        }
    )
    
    print("\n=== API Hook 測試完成 ===")

File: API-Hook/api_hook.js
/**
 * API Hook Implementation - Node.js Example
 * 用於監控與記錄 API 呼叫的 Node.js/Express 實作範例
 */

const crypto = require('crypto');
const fs = require('fs');
const yaml = require('js-yaml');

/**
 * API Hook 核心類別
 */
class APIHook {
  /**
   * 初始化 API Hook
   * @param {Object} options - 配置選項
   * @param {string} options.configPath - 配置檔案路徑
   */
  constructor(options = {}) {
    this.config = this._loadConfig(options.configPath || 'config.yaml');
    this.sensitiveFields = this.config.security?.sensitive_fields || [];
    this.logs = [];
    console.log('API Hook initialized');
  }

  /**
   * 載入配置檔案
   * @private
   */
  _loadConfig(configPath) {
    try {
      // 簡化實作：實際應從 YAML 檔案載入
      return {
        security: {
          sensitive_fields: ['password', 'api_key', 'token', 'secret'],
          encrypt_sensitive_data: true,
          hash_algorithm: 'SHA256'
        },
        logging: {
          level: 'INFO',
          storage: 'database'
        },
        performance: {
          async_processing: true
        }
      };
    } catch (error) {
      console.error('Failed to load config:', error);
      return {};
    }
  }

  /**
   * 清理敏感資料
   * @private
   */
  _sanitizeData(data) {
    if (!data || typeof data !== 'object') return data;

    const sanitized = { ...data };

    this.sensitiveFields.forEach(field => {
      if (sanitized[field]) {
        const value = String(sanitized[field]);
        const hash = crypto
          .createHash('sha256')
          .update(value)
          .digest('hex')
          .substring(0, 16);
        sanitized[field] = `[REDACTED:${hash}]`;
      }
    });

    return sanitized;
  }

  /**
   * 創建日誌條目
   * @private
   */
  _createLogEntry(options) {
    return {
      timestamp: new Date().toISOString(),
      request_id: options.requestId,
      user_id: options.userId || 'anonymous',
      source_ip: options.sourceIp || '0.0.0.0',
      method: options.method || 'GET',
      endpoint: options.endpoint,
      parameters: this._sanitizeData(options.parameters),
      response_code: options.responseCode,
      response_time_ms: Math.round(options.responseTimeMs * 100) / 100,
      security_context: {
        authentication_method: options.authMethod || 'OAuth2',
        authorization_level: options.authLevel || 'User',
        session_id: options.sessionId || options.requestId
      },
      result: options.result,
      error_message: options.errorMessage || null
    };
  }

  /**
   * 儲存日誌
   * @private
   */
  _saveLog(logEntry) {
    console.log('API Log:', JSON.stringify(logEntry));
    this.logs.push(logEntry);

    // 實際實作範例：
    // if (this.config.logging.storage === 'database') {
    //   await this.db.insert('api_logs', logEntry);
    // } else if (this.config.logging.storage === 'elasticsearch') {
    //   await this.es.index('api-logs', logEntry);
    // }
  }

  /**
   * 檢查異常行為
   * @private
   */
  _checkAnomalies(logEntry) {
    // 檢查回應時間
    if (logEntry.response_time_ms > 5000) {
      this._triggerAlert(
        'Slow Response Time',
        `API ${logEntry.endpoint} took ${logEntry.response_time_ms}ms`,
        'WARNING'
      );
    }

    // 檢查錯誤
    if (logEntry.result === 'error') {
      this._triggerAlert(
        'API Error',
        `API ${logEntry.endpoint} failed: ${logEntry.error_message}`,
        'ERROR'
      );
    }

    // 檢查未授權存取
    if (logEntry.response_code === 401) {
      this._triggerAlert(
        'Unauthorized Access',
        `Unauthorized access attempt from ${logEntry.source_ip}`,
        'CRITICAL'
      );
    }
  }

  /**
   * 觸發警報
   * @private
   */
  _triggerAlert(alertName, message, severity) {
    const alert = {
      timestamp: new Date().toISOString(),
      alert_name: alertName,
      message: message,
      severity: severity
    };

    console.warn('ALERT:', JSON.stringify(alert));

    // 實際實作：發送到警報系統
    // await this._sendWebhook(alert);
    // await this._sendEmail(alert);
  }

  /**
   * Express.js 中介軟體
   * 自動監控所有 HTTP 請求
   */
  middleware(options = {}) {
    const {
      logRequests = true,
      validateAuth = false,
      trackPerformance = true
    } = options;

    return (req, res, next) => {
      const startTime = Date.now();
      const requestId = crypto.randomUUID();

      // 儲存原始的 res.json 方法
      const originalJson = res.json.bind(res);

      // 覆寫 res.json 以攔截回應
      res.json = (body) => {
        const responseTimeMs = Date.now() - startTime;

        if (logRequests) {
          const logEntry = this._createLogEntry({
            requestId: requestId,
            userId: req.user?.id || req.session?.userId,
            sourceIp: req.ip || req.connection.remoteAddress,
            method: req.method,
            endpoint: req.path,
            parameters: {
              query: req.query,
              body: req.body,
              params: req.params
            },
            responseCode: res.statusCode,
            responseTimeMs: responseTimeMs,
            result: res.statusCode < 400 ? 'success' : 'error',
            errorMessage: body.error || null,
            authMethod: req.authMethod,
            authLevel: req.user?.role,
            sessionId: req.sessionID
          });

          this._saveLog(logEntry);
          this._checkAnomalies(logEntry);
        }

        // 呼叫原始的 json 方法
        return originalJson(body);
      };

      next();
    };
  }

  /**
   * 端點保護裝飾器
   * 為特定端點添加額外的監控與驗證
   */
  protect(options = {}) {
    const { level = 'medium', logBody = true } = options;

    return (req, res, next) => {
      // 添加額外的安全檢查
      if (level === 'critical' || level === 'high') {
        // 檢查速率限制
        // 驗證 token
        // 記錄詳細資訊
      }

      // 標記此端點的安全等級
      req.securityLevel = level;
      req.logBody = logBody;

      next();
    };
  }

  /**
   * 手動記錄事件
   */
  logEvent(eventType, severity, message, metadata = {}) {
    const event = {
      timestamp: new Date().toISOString(),
      event_type: eventType,
      severity: severity,
      message: message,
      metadata: metadata
    };

    console.log('Event:', JSON.stringify(event));
    this._saveLog(event);

    // 如果是嚴重事件，觸發警報
    if (severity === 'error' || severity === 'critical') {
      this._triggerAlert(
        eventType.charAt(0).toUpperCase() + eventType.slice(1),
        message,
        severity.toUpperCase()
      );
    }
  }

  /**
   * 取得統計資訊
   */
  getStats() {
    const stats = {
      total_logs: this.logs.length,
      success_count: this.logs.filter(l => l.result === 'success').length,
      error_count: this.logs.filter(l => l.result === 'error').length,
      avg_response_time: 0
    };

    if (this.logs.length > 0) {
      const totalTime = this.logs.reduce((sum, log) => sum + (log.response_time_ms || 0), 0);
      stats.avg_response_time = Math.round(totalTime / this.logs.length * 100) / 100;
    }

    return stats;
  }
}

// 使用範例
if (require.main === module) {
  const express = require('express');
  const app = express();

  // 初始化 API Hook
  const hook = new APIHook({
    configPath: 'config.yaml'
  });

  // 使用中介軟體
  app.use(express.json());
  app.use(hook.middleware({
    logRequests: true,
    validateAuth: true,
    trackPerformance: true
  }));

  // 範例端點 1：一般 API
  app.get('/api/v1/users', (req, res) => {
    res.json({
      status: 'success',
      users: [
        { id: 1, name: 'John Doe' },
        { id: 2, name: 'Jane Smith' }
      ]
    });
  });

  // 範例端點 2：受保護的高安全性 API
  app.post('/api/v1/users', hook.protect({ level: 'high' }), (req, res) => {
    const { username, email } = req.body;

    // 模擬創建使用者
    console.log(`Creating user: ${username}`);

    res.json({
      status: 'success',
      user_id: '12345',
      username: username,
      email: email
    });
  });

  // 範例端點 3：認證 API（關鍵安全性）
  app.post('/api/v1/auth/login', hook.protect({ level: 'critical' }), (req, res) => {
    const { username, password } = req.body;

    // 模擬認證 - TEST DATA ONLY
    if (password === 'test_wrong_password') {  // TEST DATA
      return res.status(401).json({
        status: 'error',
        error: 'Invalid credentials'
      });
    }

    res.json({
      status: 'success',
      token: 'example-jwt-token-placeholder'  // TEST DATA - Not a real JWT
    });
  });

  // 範例端點 4：觸發錯誤
  app.get('/api/v1/error', (req, res) => {
    res.status(500).json({
      status: 'error',
      error: 'Internal server error'
    });
  });

  // 統計端點
  app.get('/api/stats', (req, res) => {
    res.json(hook.getStats());
  });

  // 啟動伺服器
  const PORT = process.env.PORT || 3000;
  app.listen(PORT, () => {
    console.log(`\n=== API Hook Demo Server ===`);
    console.log(`Server running on http://localhost:${PORT}`);
    console.log(`\nTest endpoints:`);
    console.log(`  GET  http://localhost:${PORT}/api/v1/users`);
    console.log(`  POST http://localhost:${PORT}/api/v1/users`);
    console.log(`  POST http://localhost:${PORT}/api/v1/auth/login`);
    console.log(`  GET  http://localhost:${PORT}/api/v1/error`);
    console.log(`  GET  http://localhost:${PORT}/api/stats`);
    console.log(`\n測試指令範例：`);
    console.log(`  curl http://localhost:${PORT}/api/v1/users`);
    console.log(`  curl -X POST http://localhost:${PORT}/api/v1/users -H "Content-Type: application/json" -d '{"username":"john","email":"john@example.com"}'`);
  });

  // 測試手動事件記錄
  setTimeout(() => {
    console.log('\n=== 測試手動事件記錄 ===');
    hook.logEvent(
      'security',
      'critical',
      'Suspicious login pattern detected',
      {
        ip: '192.168.1.100',
        attempts: 10,
        user: 'admin'
      }
    );
  }, 2000);
}

module.exports = APIHook;
